File: src/main/java/net/reflact/engine/spells/dynamic/ProjectileEffect.java
package net.reflact.engine.spells.dynamic;

import com.google.gson.JsonObject;
import net.minestom.server.coordinate.Vec;
import net.minestom.server.entity.Entity;
import net.minestom.server.entity.EntityType;

public class ProjectileEffect implements SpellEffect {
    private double speed;
    private String type; // fireball, snowball, etc.

    @Override
    public void execute(CastContext context) {
        EntityType entityType = EntityType.FIREBALL;
        if ("snowball".equalsIgnoreCase(type)) entityType = EntityType.SNOWBALL;
        
        Entity projectile = new Entity(entityType);
        projectile.setInstance(context.getInstance(), context.getOrigin());
        
        Vec direction = context.getCaster().getPosition().direction();
        projectile.setVelocity(direction.mul(speed * 20)); // Minestom velocity is per tick?
        
        // In a real system, we would register this projectile to a ProjectileManager 
        // to handle collision -> trigger subsequent effects.
        // For now, we launch it.
    }

    @Override
    public void load(JsonObject config) {
        this.speed = config.has("speed") ? config.get("speed").getAsDouble() : 1.0;
        this.type = config.has("entity_type") ? config.get("entity_type").getAsString() : "fireball";
    }
}

File: bin/main/net/reflact/engine/spells/dynamic/ProjectileEffect.kt
package net.reflact.engine.spells.dynamic

import com.google.gson.JsonObject
import net.minestom.server.entity.Entity
import net.minestom.server.entity.EntityType

class ProjectileEffect : SpellEffect {
    private var speed: Double = 0.0
    private var type: String = "" // fireball, snowball, etc.

    override fun execute(context: CastContext) {
        var entityType = EntityType.FIREBALL
        if ("snowball".equals(type, ignoreCase = true)) entityType = EntityType.SNOWBALL

        val projectile = Entity(entityType)
        projectile.setInstance(context.instance, context.origin)

        val direction = context.caster.position.direction()
        projectile.velocity = direction.mul(speed * 20) // Minestom velocity is per tick?

        // In a real system, we would register this projectile to a ProjectileManager 
        // to handle collision -> trigger subsequent effects.
        // For now, we launch it.
    }

    override fun load(config: JsonObject) {
        this.speed = if (config.has("speed")) config.get("speed").asDouble else 1.0
        this.type = if (config.has("entity_type")) config.get("entity_type").asString else "fireball"
    }
}
